### ISSUE TITLE
exporting qwen3-0.6B using the w8a8 quantization format, a segmentation error occurs.

### ISSUE THREAD

**Issue by cxn-selfie** [pytorch/executorch#16013]

**Comment by cccclai** [comment]
Does other quantization format work or is it just 8a8w? Also which qnn version are you using?

**Comment by 32kilometre** [comment]
I tried  w4a16 and w8a16 and they both worked. The qnn sdk version I'm using is 2.37.0.250724.

**Comment by cccclai** [comment]
@cxn-selfie can you share your intention for 8a8w? For LLMs, I think 8a8w is the least optimum option, because the activation is hard to quantize and we usually want to have more bits for activation. In the meanwhile, weight is relatively easier and we can use lower bits.

We should fix the segfault though, cc: @shewu-quic @haowhsu-quic @DannyYuyang-quic @winskuo-quic

**Comment by 32kilometre** [comment]
I’d like to compare inference speeds under different quantization parameters. Do you have any insights into how these parameters affect NPU performance? On an NPU, do different bit-widths or symmetry settings trigger distinct data paths?

**Comment by cccclai** [comment]
I think at 0.6B scale, 16a4w is still probably the optimum, because it's still memory bound. But if it's smaller, like SmolLM2 135M, it might be different. Like here https://github.com/pytorch/executorch/blob/6903f280b76439d8be820da612467f0fbd909968/examples/qualcomm/oss_scripts/llama/static_llm_quant_recipe.py#L556 I remember `use_16a8w` is faster 

Anyway, I think it still worth benchmark experiments...

**Comment by 32kilometre** [comment]
Thank you for your answer.
Does this mean that for small models (such as SmolLM2 135M), the computational overhead of dequantization (restoring the 4-bit weights to a computational data type) is more significant, while for large models (qwen3-0.6b), memory limitations are the more critical influencing factor?
Additionally, I have a basic question about quantization. I understand that in some schemes (such as the so-called "8a8w," where both activation and weights are 8 bits), the inference process can be performed entirely in the integer domain, thus eliminating the need for dequantization. Is my understanding correct?

**Comment by cccclai** [comment]
> Does this mean that for small models (such as SmolLM2 135M), the computational overhead of dequantization (restoring the 4-bit weights to a computational data type) is more significant, while for large models (qwen3-0.6b), memory limitations are the more critical influencing factor?

I think it's more about the computation cost vs memory limitation, and for computation cost, it's not limited to dequantization.

> Additionally, I have a basic question about quantization. I understand that in some schemes (such as the so-called "8a8w," where both activation and weights are 8 bits), the inference process can be performed entirely in the integer domain, thus eliminating the need for dequantization. Is my understanding correct?

I think all inference process for qnn is done in integer domain (if we didn't compile with `use_fp16`). There can be dequant compute, but I think it's more about from 4bit to higher level bits. From what I know, 16a4w_block is implemented in the software level and hence there can be extra cost for the "dequant", QCOM folks might know more details here.

**Comment by Ulthunzy** [comment]
**Summary**
 I am using Executorch’s ETDump feature to analyze execution time distribution of a QNN model.
 I generated the ETDump file using the following code snippet in **_examples/qualcomm/oss_scripts/llama/qnn_llama_runner.cpp_**

```
 #include <executorch/backends/qualcomm/runtime/QnnExecuTorch.h>
 #include <executorch/examples/qualcomm/oss_scripts/llama/runner/runner.h>
 #include <executorch/extension/llm/runner/irunner.h>
 #include <executorch/runtime/platform/log.h>
 #include <gflags/gflags.h> 
 #include <fstream>
 #include <vector>
 #include <executorch/devtools/etdump/etdump_flatcc.h>

 DEFINE_string(
  etdump_path,
  "outputs/etdump.etdump",
  "If set, dump ETDump to this path.");
  ...
 
  ...
 
 template <typename T>
 void start_runner(
     std::unique_ptr<executorch::extension::Module> module,
     std::vector<std::string>& prompts,
     **executorch::etdump::ETDumpGen* etdump_raw = nullptr**) {

   ......

   if (etdump_raw) {
     auto result = etdump_raw->get_etdump_data();
     if (result.buf != nullptr && result.size > 0) {
       FILE* f = fopen(FLAGS_etdump_path.c_str(), "w+");
       fwrite((uint8_t*)result.buf, 1, result.size, f);
       fclose(f);
       free(result.buf);
     }
   }
 }
 
 int main(int argc, char** argv) {
   std::vector<std::string> prompts = CollectPrompts(argc, argv);
   gflags::ParseCommandLineFlags(&argc, &argv, true);
   if (!gflags::GetCommandLineFlagInfoOrDie("prompt").is_default &&
       !gflags::GetCommandLineFlagInfoOrDie("tokenized_prompt").is_default) {
     ET_CHECK_MSG(false, "Only provide prompt or tokenized_input but not both.");
   }
   if (!gflags::GetCommandLineFlagInfoOrDie("dump_logits_path").is_default &&
       FLAGS_eval_mode != 0) {
     ET_CHECK_MSG(
         false, "Only TokenGenerator(kv) mode is supported to dump all logits.");
   }
   auto etdump = std::make_unique<executorch::etdump::ETDumpGen>(); 
   auto* etdump_ptr = etdump.get();
   std::unique_ptr<executorch::extension::Module> module =
       std::make_unique<executorch::extension::Module>(
           FLAGS_model_path.c_str(),
           executorch::extension::Module::LoadMode::MmapUseMlockIgnoreErrors,
           std::move(etdump));
   // Using 8bit as default since this meta is introduced with 16bit kv io
   // support and older models only have 8bit kv io.
   example::KvBitWidth kv_bitwidth = example::KvBitWidth::kWidth8;
   if (module->method_names()->count("get_kv_io_bit_width") > 0) {
     kv_bitwidth = static_cast<example::KvBitWidth>(
         module->get("get_kv_io_bit_width").get().toScalar().to<int64_t>());
   }
 
   if (kv_bitwidth == example::KvBitWidth::kWidth8) {
     start_runner<uint8_t>(std::move(module), prompts, etdump_ptr);
   } else if (kv_bitwidth == example::KvBitWidth::kWidth16) {
     start_runner<uint16_t>(std::move(module), prompts, etdump_ptr);
   } else {
     ET_CHECK_MSG(
         false,
         "Unsupported kv bitwidth: %ld",
         static_cast<int64_t>(kv_bitwidth));
   }
 
   return 0;
 }

```
While analyzing the ETDump output, I noticed a large amount of copy and to_copy time.
This raised a few questions regarding where these operations are executed and whether certain ops are unexpectedly falling back to the host.

**1. Are the copy / to_copy operations executed on the device or on the host?**

The ETDump output does not clearly indicate whether these operations run on the QNN device or on the host side, so it's hard to determine if the overhead comes from QNN runtime or host-side memory movement.

**2. Conv op seems to fall back to host execution**

Based on the timing breakdown, it appears that some conv operators are not executed on the device backend and instead fall back to host execution.
This is unexpected because convolutions are typically offloaded to the QNN backend.

**My current assumption (may be incorrect)**

1. Some of the copy operations might correspond to host↔device data transfers.

2. Certain conv ops might not satisfy QNN backend requirements and therefore fall back to host execution.

However, I couldn’t find documentation explaining these behaviors in detail.

Below is a small excerpt of the ETDump trace in table format.
The full trace is available here:
<img width="1686" height="358" alt="Image" src="https://github.com/user-attachments/assets/c989e076-876f-448a-ac44-56433ce657d9" />

https://github.com/Ulthunzy/qnn-trace-data/blob/master/etdump.xlsx

This bar chart was generated from the ETDump trace data during my analysis.

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/2a696f81-e145-496c-b7f0-6e54d3a611dc" />

**Comment by shewu-quic** [comment]
Hi @Ulthunzy ,

Thanks for your detailed profiling. Let me try to reply your questions first.
> 1. Are the copy / to_copy operations executed on the device or on the host?

To clarify your terms: executing on the device refers to running on the HTP, while executing on the host refers to running on the CPU, correct?
In the etdump results, you can check the `is_delegated_op` column to determine whether the operation runs on the HTP or CPU. If `is_delegated_op` is True, the operation runs on the HTP.

Regarding [to_copy operation](https://github.com/pytorch/executorch/blob/main/backends/qualcomm/builders/op_to.py), it is used for requantization. Since we have applied mix-precision in the graph, this leads to a transformation from `x->node (u16) -> node(u8) ->...`. As a result, we insert the `to_copy` operation to cast or convert the data type.
As for the other copy operation, I'm not sure which one you are referring to. Could you please clarify your observation about the copy operation?

> 2. Conv op seems to fall back to host execution

Based on your results, none of the convolution operations will fall back to the CPU—they should all run on HTP. To confirm, you can review your export PTE stage in AOT and ensure that all convolution operations show `[QNN Partitioner Op Support]: aten.convolution.default | True`.

For more information on profiling details, please refer to this image or [QNN doc](https://docs.qualcomm.com/doc/80-63442-10/topic/htp_backend.html#qnn-htp-profiling)

<img width="997" height="775" alt="Image" src="https://github.com/user-attachments/assets/290420f8-5e6b-455e-b6bf-1e0bc6b63b65" />

To clarify your results, I have some questions:
1. Can you let me know which decoder model you use and what commands you work with?
2. What device are you using?
4. [We have applied TagQuantIO pass on LLM model](https://github.com/pytorch/executorch/blob/8af82522fc7f066a9613efe3027471befd250492/examples/qualcomm/oss_scripts/llama/llama.py#L729) to prevent Q/DQ from appearing in the graph. As a result, Q/DQ should not be present in the profiling. If they do appear, it may indicate an incorrect profiling record that could be misleading or something wrong.

**Comment by Ulthunzy** [comment]
Hi @shewu-quic 
Thanks a lot for the clarification. I will answer your questions first and provide some additional details based on my ETDump results.

> 1. About the definition of device vs host

Yes, your understanding matches mine:
1. device = HTP
2. host = CPU

And I also confirmed through the is_delegated_op column in the ETDump that this mapping is correct.

> 2. About the “other copy” operators

Besides to_copy, I also observed several operators whose names include “copy”, such as:

1. aten_slice_copy_tensor_15:OpId_3483 (cycles)
2. aten_permute_copy_default_34:OpId_3900 (cycles)

These operators show 0 cycles (or extremely close to 0) in ETDump.

> So my question is:

Are these “slice_copy” / “permute_copy” ops simply lightweight view/reshape-related ops recorded in ETDump, rather than actual memory copy operations?

Originally, I was suspecting whether to_copy serves as a Q/DQ-like step, so I want to confirm my understanding here.

> 3. Model and commands I used

The decoder model I used is:
Qwen3-0.6B

> The command for exporting the model was:

```
python examples/qualcomm/oss_scripts/llama/llama.py \
  -a ./llama-output/qwen3-0_6b-m1024-b128-kv-w8a16-256 \
  -b build-android \
  -s 431e3c37 \
  -m SM8750 \
  --temperature 0.3 \
  --model_mode kv \
  --max_seq_len 1024 \
  --prefill_ar_len 128 \
  --decoder_model qwen3-0_6b \
  --prompt "I would like to learn python, could you teach me with a simple example?" \
  --compile_only
```

> The command used for running inference was:
```
python examples/qualcomm/oss_scripts/llama/llama.py \
  -b build-android \
  -s 431e3c37 \
  -m SM8750 \
  --decoder_model qwen3-0_6b \
  --model_mode kv \
  --prefill_ar_len 128 \
  --max_seq_len 4096 \
  --tasks wikitext \
  --limit 1 \
  --prompt "what is 1+1?" \
  --pre_gen_pte ./llama-output/qwen3-0_6b-m1024-b128-kv-w8a16-256
```

> 4. About Q/DQ nodes

In my ETDump results, I did not see explicit Quantize / Dequantize or Q/DQ ops in the event_name column.
So this seems consistent with your description that TagQuantIO should remove these nodes.


> Additionally, I observed something even more confusing in the ETDump results:

Some non-trivial compute ops — for example:
_aten_sigmoid_default_3:OpId_9157 (cycles)_
also show avg(cycles) = 0, and all percentile values (p10/p50/p90) are also 0.
Since sigmoid is a real compute operator and cannot be a no-op or a lightweight view operation, I would expect it to have a measurable execution cost (either on CPU or HTP).
However, ETDump still reports 0 cycles, similar to view/slice/permute ops.

So I would like to understand:

1. Why does ETDump report 0 cycles even for compute-heavy ops like sigmoid?
2. Does this indicate that the op was fused into another operator and not executed independently?
3. Or is the profiler sampling interval too coarse to catch short CPU/HTP ops?
4. Or is ETDump only recording delegate-level blocks and not per-op timing for some operators?

Is this expected behavior for small ops in LLM workloads?

Understanding why even aten_sigmoid_default shows 0 cycles would help me better interpret the profiling results and determine whether I'm missing some timing information.

**Comment by shewu-quic** [comment]
> Besides to_copy, I also observed several operators whose names include “copy”, such as:
> aten_slice_copy_tensor_15:OpId_3483 (cycles)
> aten_permute_copy_default_34:OpId_3900 (cycles)
> Are these “slice_copy” / “permute_copy” ops simply lightweight view/reshape-related ops recorded in ETDump, rather than 
> actual memory copy operations?

[aten_permute_copy_default_34 might be generated from transpose or permute](https://github.com/pytorch/executorch/blob/2d1f181db49e452ac9eaf664c879756ab3183697/examples/qualcomm/oss_scripts/llama/model/static_llama.py#L188) and [aten_slice_copy_tensor_15 might be source from tensor slicing](https://github.com/pytorch/executorch/blob/2d1f181db49e452ac9eaf664c879756ab3183697/examples/qualcomm/oss_scripts/llama/model/apply_rope.py#L43). All of them run on HTP.

> Why does ETDump report 0 cycles even for compute-heavy ops like sigmoid?
> Does this indicate that the op was fused into another operator and not executed independently?
> Or is the profiler sampling interval too coarse to catch short CPU/HTP ops?
> Or is ETDump only recording delegate-level blocks and not per-op timing for some operators?

In detailed profiling, if the operation cycle is zero, it could [indicate either fusion in QNN HTP or parallel execution of operations](https://docs.qualcomm.com/doc/80-63442-10/topic/htp_backend.html#qnn-htp-profiling).

I am glad to help you to clarify profiling results :)
If possible, try using QHAS to profile the LLM for more detailed insights and better readability.
Please refer to [README.md](https://github.com/pytorch/executorch/tree/main/backends/qualcomm/debugger) and [set optrace = True in llama.py](https://github.com/pytorch/executorch/blob/2d1f181db49e452ac9eaf664c879756ab3183697/examples/qualcomm/oss_scripts/llama/llama.py#L379) for instructions on generating QHAS.

If you encounter any issues, please feel free to let me know.

**Comment by Ulthunzy** [comment]
Thanks for the explanation. I have one more question regarding weight precision.

> Our exported model uses w4a16, but in the HTP-level profiling I never see any operator using int4 (qint4/quint4).
> The lowest precision shown in ETDump is always qint8/quint8.

My understanding is that this is expected because:

- HTP cannot load memory in 4-bit granularity
- Weights must be stored as 8-bit containers (each byte holding two 4-bit values)
- During computation, the kernel unpacks the byte into two int4 weights
- Then the unpacked values are converted to int8 internally
- And the actual MAC is still int8 × int8 → int32

Is this interpretation correct?

> In other words:
> INT4 is only a storage format, not an execution format on HTP, so profiling naturally will not show int4 dtype.

Please correct me if I misunderstand anything.

**Comment by shewu-quic** [comment]
Currently, [there are limited some ops be supported 4-bit encoding weight in QNN HTP](https://docs.qualcomm.com/doc/80-63442-10/topic/htp_guidelines_int4_weights.html). These operations use the int4 data type weight for computation, but [the data is stored as int8 with a 4-bit bitwidth](https://github.com/pytorch/executorch/blob/main/backends/qualcomm/builders/node_visitor.py#L241).

**Comment by 32kilometre** [comment]
Hi @shewu-quic 
Thank you very much for your answers.

My analysis of a convolution operation shows a large number of `q::Slice_contig.tcm` operators during the dequantization stage. 

<img width="1507" height="703" alt="Image" src="https://github.com/user-attachments/assets/505ef30d-0094-4548-9f12-95dc44228cc7" />

My understanding is that the matrix size after the concat operation is too large to be directly stored in the TCM, it needs to be segmented before dequantization. However, I'm confused by the fact that this operator has three inputs. Could you please explain the specific function of this operator?

Please correct me if I misunderstand anything.

**Comment by shewu-quic** [comment]
Hi @cxn-selfie,
Thank you for sharing your detailed analysis! 
Your understanding of why segmentation occurs is generally on the right track—hardware resource constraints often require data to be split into smaller chunks for processing. This helps ensure efficient memory usage and performance.

Regarding the specific operator behavior and its three inputs, those details are part of the framework’s internal implementation and aren’t fully documented publicly. However, you can refer to the official operator documentation and API guides for more insight into its design principles and expected input/output behavior.

If your concern is related to performance or correctness in your model, feel free to share your configuration and observations. 
We’d be happy to discuss optimization strategies or clarify any usage questions.
